### PR TITLE
[docs] AWS database discovery

### PR DESCRIPTION
**Pull request by greedy52**
Related issues:
- #23774
- #30569

Changes:

- Added a new `AWS Discovery` guide
- Updated all existing AWS guides that support auto-discovery to use dynamic resources (from auto-discovery, tcl, Terrform) and static config, replacing "legacy" database discovery
- Updated AWS cross-account and dynamic resource guides

This is a quite big change in my opinion. **I'm up for any suggestions on how to structure this better!!!!**

Next/TODO PRs:
- Update `Maximum policy size exceeded errors`
- Azure database discovery

Note on include/partials:
Fighting this a lot as `if-else` condition is not supported. Instead
- Duplicated a few partials with suffix `-with-account-id` to support AWS databases that requires AWS account ID for setup
- Duplicated a few partials with suffix `-with-var-protocol` to support using `<Var name=""/>` for protocol instead of a string literal 

This guide is for v14 only, not backporting.

### CODE REVIEW

**Comment by ptgott** [comment]
@greedy52 I wonder if it would simplify this work if we were to:
- Include a single method of applying dynamic resources within each how-to guide, rather than showing `tctl`, Terraform, etc.
- In the "Next steps" section of each how-to guide, explain that you can use other methods of applying dynamic resources, and link to the "Using Dynamic Resources" section of the docs (https://goteleport.com/docs/management/dynamic-resources). 

The idea is that we would (a) educate users about a workflow in the how-to guides and (b) educate users about alternatives to the workflow in the "Using Dynamic Resources" section.

**Comment by greedy52** [comment]
> @greedy52 I wonder if it would simplify this work if we were to:

Removed `tctl` and Terraform from these guides. Still, the guides are very complex, especially the RDS guide with OIDC-for-kube-cluster setups. 😭

**Review by smallinsky — COMMENTED** [review]
Thanks for this docs update.
First pass. I have left comments related to docs struct but mostly seems to be reasonable.

File: docs/pages/includes/database-access/aws-register-resource-static.mdx [hunk 1]
```diff
@@ -0,0 +1,4 @@
+You can define databases when creating the Database Service configuration.
+
+Take a note of your database's URI which will be used in a later step. For
+example: <Var name="{{ example }}"/>.
```
[the remark this answers is not in the document]
Comment (greedy52) [reply docs/pages/includes/database-access/aws-register-resource-static.mdx:4]: I didn't want to move `teleport db configure create` to step 1. Let me try to merge `Register the database` and `Create a db configuration` in one step.

**Comment by ptgott** [comment]
Hi @greedy52, looks like there are a few conflicts here. I'll give this a look once these are fixed!

File: docs/pages/includes/database-access/discovery-guide-brief.mdx [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+The first, the Discovery Service, is responsible for watching your cloud
+provider and checking if there are any new databases or if there have been any
+modifications to previously discovered databases.
+
+The second, the Database Service, monitors the databases created by the
+Discovery Service and proxies communications between users and the database
+servers hosted by the cloud provider.
+
+In this guide, you will configure a Teleport Discovery Service to discover
```
Comment (ptgott) [inline docs/pages/includes/database-access/discovery-guide-brief.mdx:11]:
```suggestion
In this guide, you will configure the Teleport Discovery Service to discover
```
or "a Teleport Discovery Service instance"

File: docs/pages/includes/database-access/discovery-guide-brief.mdx [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+{{ cloud }}-hosted databases automatically.
+
+<Admonition type="note">
```
Comment (ptgott) [inline docs/pages/includes/database-access/discovery-guide-brief.mdx:14]: I think we can remove this admonition, since it's not relevant for a user who is on v14 of the docs and following the guides that include this partial.

For a user on a previous version of the docs, they'll see the legacy instructions anyway.

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+title: AWS Database Auto-Discovery
+description: How to configure Teleport to discover AWS-hosted databases.
+---
+
+(!docs/pages/includes/database-access/discovery-guide-brief.mdx cloud="AWS" !)
+
+<Admonition type="tip">
+
+Discovery Service can run independently on a different machine from Database
+Services.
```
Comment (ptgott) [inline docs/pages/database-access/guides/aws-discovery.mdx:11]:
```suggestion
The Discovery Service can run independently on a different machine from Database
Service instances.
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 5]
```diff
@@ -0,0 +1,176 @@
+---
+title: AWS Database Auto-Discovery
+description: How to configure Teleport to discover AWS-hosted databases.
+---
+
+(!docs/pages/includes/database-access/discovery-guide-brief.mdx cloud="AWS" !)
+
+<Admonition type="tip">
```
Comment (ptgott) [inline docs/pages/database-access/guides/aws-discovery.mdx:8]: I would incorporate this information into the Prerequisites, since otherwise it strays from the focus of this guide a bit.

Maybe we could recommend either one host for both the Discovery Service and the Database Service or separate hosts for each.

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+(!docs/pages/includes/edition-prereqs-tabs.mdx!)
+
+- An AWS account with permissions to create and attach IAM policies.
+- One or more database servers hosted by AWS database services.
+- A host to run the Teleport Discovery Service.
+
+## Step 1/4. Generate a join token
+
+The Discovery Service require a valid auth token to connect to the cluster.
```
Comment (ptgott) [inline docs/pages/database-access/guides/aws-discovery.mdx:32]: I think we should replace "auth token" with "join token" to be consistent with the heading

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+Once the Discovery Service starts, database servers matching the tags and
+regions specified in the AWS section will be added to the Teleport cluster
+automatically.
+
+You can list them with the `tctl get db` command or check a specific database
+with `tctl get db/<database-name>`.
+
+## Next
+- Learn [Dynamic Registration](./dynamic-registration.mdx) by Database Service.
```
Comment (ptgott) [inline docs/pages/database-access/guides/aws-discovery.mdx:174]:
```suggestion
- Learn about [Dynamic Registration](./dynamic-registration.mdx) by the Teleport Database Service.
```

File: docs/pages/database-access/guides/dynamic-registration.mdx [hunk 8]
```diff
@@ -104,3 +119,8 @@ To delete a database resource, run:
 ```code
 $ tctl rm db/example
 ```
+
+Aside from `tctl`, dynamic resources can also be added by:
```
Comment (ptgott) [inline docs/pages/database-access/guides/dynamic-registration.mdx:123]: Maybe we could link to the "Dynamic Resources" subsection of the docs here?

File: docs/pages/database-access/guides/redis-aws.mdx [hunk 9]
```diff
@@ -37,73 +38,26 @@ description: How to configure Teleport database access with AWS ElastiCache and
 
 (!docs/pages/includes/database-access/alternative-methods-join.mdx!)
 
-Install Teleport on the host where you will run the Teleport Database Service:
+<Tabs>
 
-(!docs/pages/includes/install-linux.mdx!)
+<TabsItem label="ElastiCache">
```
Comment (ptgott) [inline docs/pages/database-access/guides/redis-aws.mdx:43]: Should be `TabItem` (there are other instances in this PR as well)

File: docs/pages/includes/database-access/aws-create-db-configuration-modify-with-account-id.mdx [hunk 10]
```diff
@@ -0,0 +1,33 @@
+Modify your Teleport Database Service static configuration file:
```
Comment (ptgott) [inline docs/pages/includes/database-access/aws-create-db-configuration-modify-with-account-id.mdx:1]:
```suggestion
Modify your Teleport Database Service static configuration file:

```

File: docs/pages/includes/database-access/aws-create-db-configuration-modify.mdx [hunk 11]
```diff
@@ -0,0 +1,31 @@
+Modify your Teleport Database Service static configuration file:
```
Comment (ptgott) [inline docs/pages/includes/database-access/aws-create-db-configuration-modify.mdx:1]:
```suggestion
Modify your Teleport Database Service static configuration file:

```

File: docs/pages/includes/database-access/aws-create-db-configuration-dynamic-resources.mdx [hunk 12]
```diff
@@ -0,0 +1,30 @@
+Databases can be registered dynamically by Discovery Service, `tctl`, etc.
+
+Generate a Database Service configuration that monitors the dynamic database
+resources:
+
+<Tabs>
+<TabItem scope={["oss", "enterprise"]} label="Self-Hosted">
```
Comment (ptgott) [inline docs/pages/includes/database-access/aws-create-db-configuration-dynamic-resources.mdx:7]: The difference between tab items is minimal here, so I would recommend using a `Var` or the Cloud URL for both. (Same with the other partial that includes `Tabs` like this)

File: docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx [hunk 13]
```diff
@@ -0,0 +1,27 @@
+{{ partialSuffix="" }}
+
+Install Teleport on the host where you will run the Teleport Database Service:
+
+(!docs/pages/includes/install-linux.mdx!)
+
+On the node that is running the Database Service, create a configuration file
```
Comment (ptgott) [inline docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx:7]: I think it would be a lot simpler to recommend one method of creating a configuration and guiding readers toward the "Dynamic Resources" docs subsection intro at the end of each page.

The way I see it, each how-to guide should include a small number of paths that will work for most users (or that most users can follow as an example for their own use case), then direct users to guides that let them pick a more customized approach.

File: docs/pages/includes/database-access/aws-create-db-configuration.mdx [hunk 14]
```diff
@@ -0,0 +1,26 @@
+{{ partialSuffix="" }}
+
+Install Teleport on the host where you will run the Teleport Database Service:
+
+(!docs/pages/includes/install-linux.mdx!)
+
+On the node that is running the Database Service, create a configuration file
+depending how the database is registered:
+
+<Tabs>
```
Comment (ptgott) [inline docs/pages/includes/database-access/aws-create-db-configuration.mdx:10]: Same comment as above re: using a single approach here.

**Review by ptgott — COMMENTED** [review]
Let's make sure all pages render properly in `yarn dev`. At least one page includes a `TabsItem` where it should be a `TabItem`.

I kept my review pretty light on the language, assuming most of these changes are refactors.

File: docs/pages/database-access/guides/aws-discovery.mdx (hunk 5 above)
Comment (greedy52) [reply docs/pages/database-access/guides/aws-discovery.mdx:8]: I will just remove the tip. I think the current `Prerequisites` implies a separate host for discovery service anyway, which should be the preferred configuration.

File: docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx (hunk 13 above)
Comment (greedy52) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx:7]: > The way I see it, each how-to guide should include a small number of paths that will work for most users 

Agreed on small number of paths. However we do get ppl asking how to update existing config files or how to configure static databases. They may not be the majority cases but still generate some support load. So it's important to document them.

Doing these tabs is the easiest way I can think of. If we are to drop the other tabs here, what would you recommend so ppl on those paths can easily figure out what to do?

Thanks!

File: docs/pages/database-access/guides/dynamic-registration.mdx (hunk 8 above)
Comment (ptgott) [reply docs/pages/database-access/guides/dynamic-registration.mdx:123]: That's right, since it's a conceptual overview of dynamic resources.

File: docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx (hunk 13 above)
Comment (ptgott) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx:7]: What would you think of leaving only the "Modifying existing Database Service" content? I like the idea of showing all of the relevant configuration options within the guide as a way of educating users about the configuration file. It's also faster than the other two approaches, since it doesn't require running a binary, and users can copy the config into a user data script as shown in the [Deploy Teleport Agents with Terraform](https://goteleport.com/docs/agents/deploy-agents-terraform/) guide.

The "Next steps" section links to the docs on dynamic registration, so users should eventually learn about this technique for registering databases.

**Review by ptgott — APPROVED** [review]
Just one outstanding item re: a `Tabs` component ([this comment](https://github.com/gravitational/teleport/pull/30913#discussion_r1319901633)). Approving since this will be straightforward to act on if you'd like!

**Comment by greedy52** [comment]
> Just one outstanding item re: a `Tabs` component ([this comment](https://github.com/gravitational/teleport/pull/30913#discussion_r1319901633)). Approving since this will be straightforward to act on if you'd like!

It didn't end up an easy change:

- I need to add other parts of the `teleport.yaml` so it can be used directly by `teleport start`
- Changed to the code block with `sudo cat <<EOF`, but it doesn't allow new lines between yaml blocks and the yaml looks pretty bad in my opinion.

Let me know what you think about the last change.

**Comment by ptgott** [comment]
@greedy52 You should be able to add `Var`s to the `yaml` blocks without making them `code` blocks! Let me know if there are any issues with this, since that would now be a bug in the docs engine.

File: docs/pages/includes/database-access/aws-create-db-configuration-with-account-id.mdx [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      # Optional AWS role that the Database Service will assume to access
+      # this database.
+      assume_role_arn: ""
+      # Optional AWS external ID that the Database Service will use to assume
+      # a role in an external AWS account.
+      external_id: ""
+    static_labels:
+      env: prod
+EOF
+```
```
Comment (smallinsky) [inline docs/pages/includes/database-access/aws-create-db-configuration-with-account-id.mdx:51]: This reveals a configuration  lot of complexity. 
Any way to use `teleport db configure`  instead  or hide this behind somthing like "Teleport Manual configuration file" ?

File: docs/pages/includes/database-access/aws-create-db-configuration.mdx [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+      # Optional AWS role that the Database Service will assume to access
+      # this database.
+      assume_role_arn: ""
+      # Optional AWS external ID that the Database Service will use to assume
+      # a role in an external AWS account.
+      external_id: ""
+    static_labels:
+      env: prod
+EOF
+```
```
Comment (smallinsky) [inline docs/pages/includes/database-access/aws-create-db-configuration.mdx:50]: Same remark as  in docs/pages/includes/database-access/aws-create-db-configuration-with-account-id.mdx

**Review by smallinsky — COMMENTED** [review]
Mostly LGTM. Left one remake about static teleport db configuration file approach.

File: docs/pages/includes/database-access/aws-create-db-configuration-with-account-id.mdx (hunk 15 above)
Comment (greedy52) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-account-id.mdx:51]: @smallinsky this was done to avoid needing multiple tabs on creating configurations for different situations (dynamic resources vs static vs modifying existing config). I think it's either this, or `teleport db configure` for dynamic resources. What do you think?

Previous https://github.com/gravitational/teleport/pull/30913#discussion_r1316380560

File: docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx (hunk 13 above)
Comment (smallinsky) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx]: But the  `teleport db configure create`  is our  recommended way to configure db service. 
By covering only  "Static Configuration" approach we will not guide to use `teleport db configure create` command and also will also introduce inconsistency with Self-hosted DB guides  where we guide users to use `teleport db configure create` command. 

I agree that showing static configuration is very valuable but I think that we should do that in addition to our default/recommended `teleport db configure create` flow.

Comment (ptgott) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx]: Showing the `teleport db configure create` command works for me! We could then include a paragraph/bullet point in the "Next steps" section that briefly tells the user how to edit an existing config file, pointing to the config reference and the "Teleport Agents" [section intro](https://goteleport.com/docs/agents/introduction/).

My main goal here is to have a single path for configuring the Database Service that satisfies 80% of users to avoid a ton of complexity in this guide. The other 20% can read our configuration reference etc.

Comment (greedy52) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx]: If we only show a single path with `teleport db configure create`, then it will be for dynamic resources. I will make this change if we agree. Probably there is no single path that will cover 80% of users. 

> We could then include a paragraph/bullet point in the "Next steps" section that briefly tells the user how to edit an existing config file, pointing to the config reference and the "Teleport Agents" [section intro](https://goteleport.com/docs/agents/introduction/).

"Next Steps" already links to the YAML reference. Maybe we can reword it to something like `To understand or update the YAML configuration, refer to the configuration reference.` ?

Comment (ptgott) [reply docs/pages/includes/database-access/aws-create-db-configuration-with-var-protocol.mdx]: That sounds good to me!

**Review by ptgott — APPROVED** [review]

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+(!docs/pages/includes/edition-prereqs-tabs.mdx!)
+
+- An AWS account with permissions to create and attach IAM policies.
+- One or more database servers hosted by AWS database services.
+- A host to run the Teleport Discovery Service.
+
+## Step 1/4. Generate a join token
+
+The Discovery Service require a valid join token to connect to the cluster.
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:18]:
```suggestion
The Discovery Service requires a valid join token to connect to the cluster.
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+(!docs/pages/includes/database-access/alternative-methods-join.mdx!)
+
+## Step 2/4. Configure the Discovery service
+
+Enabling AWS database discovery requires that the `discovery_service.aws`
+section includes at least one entry and that `discovery_service.aws.types`
+includes one of database types listed in the sample YAML below. 
+
+Now create the `teleport.yaml` file on the host:
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:36]:
```suggestion
Create a `teleport.yaml` file similar to the following on the host that will run the Discovery Service:
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+## Step 3/4. Bootstrap IAM permissions
+
+Create an IAM role and attach it to the host that will run the Discovery
+Service.
+
+Teleport can bootstrap IAM permissions for the Discovery Service using the
+`teleport discovery bootstrap` command. You can use this command in automatic
+or manual mode:
+
+- In automatic mode, Teleport will attempt to create appropriate IAM policies
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:94]:
```suggestion
- In automatic mode, Teleport attempts to create the appropriate IAM policies
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  and attach them to the specified IAM roles. This requires IAM permissions to
+  create and attach IAM policies.
+- In manual mode, Teleport will print required IAM policies. You can then create
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:97]:
```suggestion
- In manual mode, Teleport prints the required IAM policies. You can then create
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+  $ teleport discovery bootstrap \
+    --attach-to-role arn:aws:iam::<Var name="aws-account-id"/>:role/<Var name="iam-role-name"/> \
+    --policy-name TeleportDatabaseDiscovery \
+    -c <Var name="/etc/teleport.yaml"/>
+  ```
+  </TabItem>
+
+  <TabItem label="Manual">
+  Use this command to display required IAM policies which you will then create
+  in your AWS console:
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:119]:
```suggestion
  Use the following command to display the required IAM policies that you need to create
  in your AWS console:
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+    --policy-name TeleportDatabaseDiscovery \
+    -c <Var name="/etc/teleport.yaml"/>
+  ```
+  </TabItem>
+</Tabs>
+
+<Details title="Bootstrapping with assume_role_arn in config">
+When `assume_role_arn` is configured for AWS matchers, `teleport discovery
+bootstrap` command will determine permissions required for the bootstrap target
+AWS IAM identity using the following logic:
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:133]:
```suggestion
bootstrap` command determines the permissions required for the bootstrap target
AWS IAM identity using the following logic:
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+
+- When the target does not match `assume_role_arn` in any AWS matcher in the
+  configuration file, the target is assumed to be the Teleport Discovery
+  Service's AWS IAM identity and permissions are bootstrapped for all the AWS
+  matchers without `assume_role_arn`.
+- When an `--attach-to-role` target matches an `assume_role_arn` setting for
+  AWS matchers in the configuration file, permissions will be bootstrapped only
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:140]:
```suggestion
  AWS matchers in the configuration file, permissions are bootstrapped only
```

File: docs/pages/database-access/guides/aws-discovery.mdx [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+IAM role that is used for `assume_role_arn`.
+</Details>
+
+## Step 3/3. Start the Discovery Service
+
+(!docs/pages/includes/start-teleport.mdx service="the Discovery Service"!)
+
+Once the Discovery Service starts, database servers matching the tags and
+regions specified in the AWS section will be added to the Teleport cluster
+automatically.
```
Comment (lsgunn-teleport) [inline docs/pages/database-access/guides/aws-discovery.mdx:154]:
```suggestion
After the Discovery Service starts, database servers matching the tags and
regions specified in the AWS section are added to the Teleport cluster
automatically.
```

**Review by lsgunn-teleport — APPROVED** [review]
A few minor wording suggestions.
